### PR TITLE
[MPS] Add native im2col

### PR DESCRIPTION
**Pull request by malfet**
It's called from `torch.unfold` and one of the few remaining vestiges in `MPSFallback.mm`

Strongly inspired by CUDA implementation from https://github.com/pytorch/pytorch/blob/09519eb1959978af7f2f909acf768485772596ab/aten/src/ATen/native/cuda/im2col.cuh#L40-L61

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Im2Col.mm [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+      std::array<int32_t, 4> padding_stride = {static_cast<int32_t>(pad_width),
+                                               static_cast<int32_t>(pad_height),
+                                               static_cast<int32_t>(stride_width),
+                                               static_cast<int32_t>(stride_height)};
+      std::array<int64_t, 4> input_sizes = {input_width, input_height, n_input_plane, batch_size};
+      std::array<int64_t, 4> input_strides = {input.stride(3), input.stride(2), input.stride(1), input.stride(0)};
+      std::array<int64_t, 4> output_strides = {output.stride(2), output.stride(1), output.stride(0), output_width};
+      getMPSProfiler().beginProfileKernel(im2colPSO, "im2col", {input});
+
+#if _CAPTURE_KERNEL
```
[the remark this answers is not in the document]
Comment (malfet) [reply aten/src/ATen/native/mps/operations/Im2Col.mm]: Removed the code

File: aten/src/ATen/native/mps/operations/Im2Col.mm [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      if (getMPSProfiler().isCaptureEnabled()) {
+        getMPSProfiler().startCapture("im2col", stream);
+      }
+#endif
```
[the remark this answers is not in the document]
Comment (malfet) [reply aten/src/ATen/native/mps/operations/Im2Col.mm]: Those are pretty heavy, and kind of unreliable, let me remove it from the final code

File: aten/src/ATen/native/mps/operations/Im2Col.mm [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+    constant ulong4          & output_strides  [[buffer(5)]],              \
+    constant long4           & input_sizes     [[buffer(6)]],              \
+    uint3                      thread_index  [[thread_position_in_grid]])
+
+INSTANTIATE_IM2COL(bool);
+INSTANTIATE_IM2COL(float);
+INSTANTIATE_IM2COL(float2);
+INSTANTIATE_IM2COL(half);
+INSTANTIATE_IM2COL(half2);
+#if __METAL_VERSION__ >= 310
```
[the remark this answers is not in the document]
Comment (malfet) [reply aten/src/ATen/native/mps/operations/Im2Col.mm:85]: Yes, see 
https://github.com/pytorch/pytorch/blob/835e7bb077f12ed0ea9b332d78f6bc3b976449eb/aten/src/ATen/native/mps/OperationUtils.mm#L59-L60
Which will be triggered from here
https://github.com/pytorch/pytorch/blob/835e7bb077f12ed0ea9b332d78f6bc3b976449eb/aten/src/ATen/native/mps/OperationUtils.mm#L215-L217

File: aten/src/ATen/native/mps/operations/Im2Col.mm [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+                                                static_cast<int32_t>(dilation_width),
+                                                static_cast<int32_t>(dilation_height)};
+      std::array<int32_t, 4> padding_stride = {static_cast<int32_t>(pad_width),
+                                               static_cast<int32_t>(pad_height),
+                                               static_cast<int32_t>(stride_width),
+                                               static_cast<int32_t>(stride_height)};
+      std::array<int64_t, 4> input_sizes = {input_width, input_height, n_input_plane, batch_size};
+      std::array<int64_t, 4> input_strides = {input.stride(3), input.stride(2), input.stride(1), input.stride(0)};
+      std::array<int64_t, 4> output_strides = {output.stride(2), output.stride(1), output.stride(0), output_width};
+      getMPSProfiler().beginProfileKernel(im2colPSO, "im2col", {input});
```
[the remark this answers is not in the document]
Comment (malfet) [reply aten/src/ATen/native/mps/operations/Im2Col.mm]: Yeah, it should capture both (though again, whole profile story is a bit confusing to me at the moment)

**Review by albanD — APPROVED** [review]
Sounds good !

**Pull request closed without merging**